src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java

```diff
@@ -118,6 +123,23 @@
     /** The angle at the start of the edge geometry. Internal representation like that of inAngle. */
     private byte outAngle;
 
+    /**
+     * The set of turn restrictions of this edge. Since most instances don't have any, we reuse
+     * a global instance in order to conserve memory.
+     *
+     * This field is optimized for low memory consumption and fast access, but modification is
+     * synchronized since it can happen concurrently.
+     *
+     * Why not use null to represent no turn restrictions?
+     * This would mean that the access would also need to be synchronized but since that is a very
+     * hot code path, it needs to be fast.
+     *
+     * Why not use a concurrent collection?
+     * That would mean that every StreetEdge has its own empty instance which would increase
+     * memory significantly.
+     */
+    private Set<TurnRestriction> turnRestrictions = NO_TURN_RESTRICTIONS;
```

Comment:
```suggestion
    private Set<TurnRestriction> turnRestrictions = Set.of();
```
`NO_TURN_RESTRICTIONS` makes me wonder if that is a special none empty set, so for collections I think we should just use the simplest thing. A immutable empty set. The `Set.of()` always return the same set, so no extra mem is accocated.

Reply:
We 6 000 000 StreetEdges in the Norway graph, du you know how many of those who have turn restrictions. Adding a ref here uses ~ 24 MB of memory.

Reply:
If my memory serves me correctly, there is about 2000 of these in the Norwegai grph, so 0,03% of the street edges have one. However the fetching of turn restrictions is around 8-10% of the traverse method of the StreetEdge, which seemed to be quite a lot

Reply:
I thought that Set.of() always returns the same instance but I couldn't actually find any documentation that this is guaranteed. Is it?

Reply:
Ah, I forgot to add `static`!

Reply:
@t2gran While I'm sure you're correct I cannot find anything about the same instance being returned on https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Set.html#of(java.util.Collection).

Reply:
@t2gran Are you aware that `NO_TURN_RESTRICTIONS` is a static final variable so should not actually use any extra memory.

Reply:
Well, I ended up using Guava's `ImmutableSet` as I couldn't get kryo to serialize `Set.copyOf`. Kryo 5 has a serializer for it we're kinda stuck at kryo 4.